### PR TITLE
Re-implementing Stage/Messaging Metrics as Pipeline Extension

### PR DESCRIPTION
**Pull request by FieldFlux**
* I've added back localActorCount back via Stage and execution since I didn't see an easy way to expose this information from an extension.

* Created a new Pipeline Extension for measuring Message Rates. It logs number of messages sent and received from a node along with rates of each message type in both directions. (Allows for monitoring error rates etc) 

* Fixed an issue in automatic metric registration where inherited metrics were not being registered.

### CODE REVIEW

**Comment by JoeHegarty** [comment]
This looks good. I'll let @DanielSperry take a look but I'm happy to merge it as long as he is.

File: actors/extensions/metrics/src/main/java/com/ea/orbit/actors/extensions/metrics/MetricsPipelineExtension.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+            if (messageMeter != null)
+                messageMeter.mark();
+        }
+
+        return super.write(ctx, message);
+    }
+
+    @ExportMetric(name="messagesReceived")
+    public int getMessagesRecievedCount()
```
Comment (DanielSperry) [inline actors/extensions/metrics/src/main/java/com/ea/orbit/actors/extensions/metrics/MetricsPipelineExtension.java]: typo?

File: actors/extensions/metrics/src/main/java/com/ea/orbit/actors/extensions/metrics/MetricsPipelineExtension.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+            Message msg = (Message) message;
+            final int messageType = msg.getMessageType();
+            Meter messageMeter = sendMeters.get(messageType);
+
+            if (messageMeter != null)
+                messageMeter.mark();
+        }
+
+        return super.write(ctx, message);
```
Comment (DanielSperry) [inline actors/extensions/metrics/src/main/java/com/ea/orbit/actors/extensions/metrics/MetricsPipelineExtension.java]: use return ctx.write(message), looks better in the stack traces.

File: actors/stage/src/main/java/com/ea/orbit/actors/Stage.java [hunk 3]
```diff
@@ -1225,4 +1226,16 @@ public String toString()
...
                 '}';
     }
+
+    @ExportMetric(name="localActorCount")
+    public int getLocalActorCount()
+    {
+        int value = 0;
+        if (execution != null)
+        {
+            value = execution.getObjectCount();
```
Comment (DanielSperry) [inline actors/stage/src/main/java/com/ea/orbit/actors/Stage.java]: actually the stage owns the `objects`, execution only has the pointer

File: actors/stage/src/main/java/com/ea/orbit/actors/runtime/Execution.java [hunk 4]
```diff
@@ -164,4 +164,9 @@ public void setObjects(final LocalObjects objects)
         this.objects = objects;
     }
 
+    public int getObjectCount()
```
Comment (DanielSperry) [inline actors/stage/src/main/java/com/ea/orbit/actors/runtime/Execution.java]: not necessary. stage has the `objects`.

File: utils/metrics/metrics/src/main/java/com/ea/orbit/metrics/MetricsManager.java [hunk 5]
```diff
@@ -256,7 +256,7 @@ private Metric buildGaugeForField(Field field, Object obj)
 
         if (obj != null)
         {
-            for (Field field : obj.getClass().getDeclaredFields())
+            for (Field field : obj.getClass().getFields())
```
Comment (DanielSperry) [inline utils/metrics/metrics/src/main/java/com/ea/orbit/metrics/MetricsManager.java]: is this ok? it won't get private, default, and protected fields

**Comment by DanielSperry** [comment]
pretty good.

I've pointed some minor things. 
@FieldFlux, If you don't have the time to fix them, we can merge tomorrow and fix them later.

**Comment by johnou** [comment]
It would be nice if you could expose the stage metrics another way eg. via the pipeline. I do not use the orbit metrics package so when compiling there is always warnings about ExportMetric missing.

File: actors/extensions/metrics/src/main/java/com/ea/orbit/actors/extensions/metrics/MetricsPipelineExtension.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    {
+        if (message instanceof Message)
+        {
+            messagesReceived.increment();
+
+            Message msg = (Message) message;
+            final int messageType = msg.getMessageType();
+            Meter messageMeter = receiveMeters.get(messageType);
+
+            if (messageMeter != null)
```
Comment (johnou) [inline actors/extensions/metrics/src/main/java/com/ea/orbit/actors/extensions/metrics/MetricsPipelineExtension.java]: missing braces?

File: utils/metrics/metrics/src/main/java/com/ea/orbit/metrics/MetricsManager.java (hunk 5 above)
Comment (FieldFlux) [inline utils/metrics/metrics/src/main/java/com/ea/orbit/metrics/MetricsManager.java]: Thats ok for now. We didn't register inaccessible fields anyway. I plan on looking at this for a future change.

**Comment by DanielSperry** [comment]
Thanks @FieldFlux !

**Pull request merged**